### ISSUE TITLE
Changing Doctrine and Route annotations to PHP attributes

### ISSUE THREAD

**Issue by NotionCommotion** [rectorphp/rector#6558]
# Question

Just learning about rectorphp and wish I had started long ago.  When running a whole set of rules (i.e. `$containerConfigurator->import(SetList::DEAD_CODE);`), I get the changes, but when executing a single rule (i.e. AnnotationToAttributeRector, DoctrineAnnotationClassToAttributeRector, TypedPropertyRector), rectorphp doesn't return any results.  Using PHP 8.0.8.  I created a [demo](https://getrector.org/demo/1ebde7d4-b072-6b48-97a2-e9a64895aced).   Any suggestion what I am doing wrong?  Thanks!

```
<?php

declare(strict_types=1);

use Rector\Core\Configuration\Option;
use Rector\Php74\Rector\Property\TypedPropertyRector;
use Rector\Set\ValueObject\SetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Rector\Php80\Rector\Class_\DoctrineAnnotationClassToAttributeRector;

use Rector\Php80\Rector\Class_\AnnotationToAttributeRector;
use Rector\Php80\ValueObject\AnnotationToAttribute;
use Symplify\SymfonyPhpConfig\ValueObjectInliner;

use Rector\Arguments\Rector\FuncCall\FunctionArgumentDefaultValueReplacerRector;
use Rector\Arguments\ValueObject\ReplaceFuncCallArgumentDefaultValue;

return static function (ContainerConfigurator $containerConfigurator): void {
    // get parameters
    $parameters = $containerConfigurator->parameters();

    // paths to refactor; solid alternative to CLI arguments
    // $parameters->set(Option::PATHS, [__DIR__ . '/src', __DIR__ . '/tests']);

    // is your PHP version different from the one your refactor to? [default: your PHP version], uses PHP_VERSION_ID format
    // $parameters->set(Option::PHP_VERSION_FEATURES, PhpVersion::PHP_72);

    // Path to phpstan with extensions, that PHPSTan in Rector uses to determine types
    $parameters->set(Option::PHPSTAN_FOR_RECTOR_PATH, getcwd() . '/phpstan-for-config.neon');

    // here we can define, what sets of rules will be applied
    // tip: use "SetList" class to autocomplete sets
    $containerConfigurator->import(SetList::CODE_QUALITY);
    $containerConfigurator->import(SetList::PHP_80);
    $containerConfigurator->import(SetList::PHP_74);
    $containerConfigurator->import(SetList::DEAD_CODE);
    $containerConfigurator->import(SetList::TYPE_DECLARATION);

    // register single rule
    $services = $containerConfigurator->services();

    $services->set(TypedPropertyRector::class)
    ->call('configure', [[
        TypedPropertyRector::CLASS_LIKE_TYPE_ONLY => false,
    ]]);

    $services->set(DoctrineAnnotationClassToAttributeRector::class)
    ->call('configure', [[
        DoctrineAnnotationClassToAttributeRector::REMOVE_ANNOTATIONS => true,
    ]]);

    $services->set(AnnotationToAttributeRector::class)
    ->call('configure', [[
        AnnotationToAttributeRector::ANNOTATION_TO_ATTRIBUTE => ValueObjectInliner::inline([
            new AnnotationToAttribute('Symfony\Component\Routing\Annotation\Route', null),
        ]),
    ]]);

    $services->set(FunctionArgumentDefaultValueReplacerRector::class)
        ->call('configure', [[
            FunctionArgumentDefaultValueReplacerRector::REPLACED_ARGUMENTS => ValueObjectInliner::inline([
                new ReplaceFuncCallArgumentDefaultValue('version_compare', 2, 'gte', 'ge'),
            ]),
        ]]);
};
```
```
<?php

namespace App;

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Annotation\Groups;
use Symfony\Component\Serializer\Annotation\SerializedName;

/**
 * @ORM\Entity()
 */
class SomeClass extends AbstractController
{
    /**
     * @ORM\Id
     * @ORM\GeneratedValue
     * @ORM\Column(type="integer")
     */
    private $id;

    /**
     * @ORM\Column(type="string", length=255)
     * @Groups({"foo:read", "foo:write"})
     * @SerializedName("bla")
     */
    private $name;

    private $type;

    /**
     * @var int
     */
    private $count;

    /**
     * @var int
     */
    private $countNotUsed;

    public function getId(): ?int
    {
        version_compare(PHP_VERSION, '5.6', 'gte');
        return $this->id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(string $name): self
    {
        $this->name = $name;

        return $this;
    }

    #[Route('/my', name: 'my')]
    public function index(): Response
    {
        return $this->json();
    }

    /**
     * @Route("/path", name="action")
     */
    public function action($count2)
    {
        $count2 = $count2 + 5;
        $this->newCount = 123;
        return $this->count * $count2;
    }

    public function testSwitch(int $i):string
    {
        switch($i) {
            case 0:
                $index = 'string'.$i;
                break;
            case 1:
                $index = 'string'.$i;
                break;
            default:
                throw new \Exception('hello');
        }
        // results in NotImplementedYetException
        //return sprintf('asdf %s %d', $index, $this->bla);
        return $index;
    }

    private function unusedFunction()
    {
        return false;
    }
}

```

```
[michael@devserver test]$ vendor/bin/rector process src --dry-run
 3/3 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
1 file with changes
===================

1) src/SomeClass.php

    ---------- begin diff ----------
--- Original
+++ New
@@ @@
 class SomeClass extends AbstractController
 {
     /**
+     * @var int
+     */
+    public $newCount;
+    /**
      * @ORM\Id
      * @ORM\GeneratedValue
      * @ORM\Column(type="integer")
+     * @var int
      */
     private $id;

@@ @@
      * @ORM\Column(type="string", length=255)
      * @Groups({"foo:read", "foo:write"})
      * @SerializedName("bla")
+     * @var string|null
      */
     private $name;

-    private $type;
-
     /**
      * @var int
      */
     private $count;

-    /**
-     * @var int
-     */
-    private $countNotUsed;
-
     public function getId(): ?int
     {
-        version_compare(PHP_VERSION, '5.6', 'gte');
+        version_compare(PHP_VERSION, '5.6', 'ge');
         return $this->id;
     }

@@ @@
     }

     #[Route('/my', name: 'my')]
-    public function index(): Response
+    public function index(): \Symfony\Component\HttpFoundation\JsonResponse
     {
         return $this->json();
     }
@@ @@

     /**
      * @Route("/path", name="action")
+     * @return float|int
      */
     public function action($count2)
     {
-        $count2 = $count2 + 5;
+        $count2 += 5;
         $this->newCount = 123;
         return $this->count * $count2;
     }
@@ @@

     public function testSwitch(int $i):string
     {
-        switch($i) {
-            case 0:
-                $index = 'string'.$i;
-                break;
-            case 1:
-                $index = 'string'.$i;
-                break;
-            default:
-                throw new \Exception('hello');
-        }
+        $index = match ($i) {
+            0 => 'string'.$i,
+            1 => 'string'.$i,
+            default => throw new \Exception('hello'),
+        };
         // results in NotImplementedYetException
         //return sprintf('asdf %s %d', $index, $this->bla);
         return $index;
-    }
-
-    private function unusedFunction()
-    {
-        return false;
     }
 }
    ----------- end diff -----------

Applied rules:
 * FunctionArgumentDefaultValueReplacerRector (https://php.watch/versions/8.1/version_compare-operator-restrictions)
 * CombinedAssignRector
 * CompleteDynamicPropertiesRector
 * RemoveUnusedPrivateMethodRector
 * RemoveUnusedPrivatePropertyRector
 * ChangeSwitchToMatchRector (https://wiki.php.net/rfc/match_expression_v2)
 * AddArrayReturnDocTypeRector
 * ReturnTypeDeclarationRector (https://wiki.php.net/rfc/scalar_type_hints_v5)
 * PropertyTypeDeclarationRector



 [OK] 1 file would have changed (dry-run) by Rector


[michael@devserver test]$

```

**Comment by TomasVotruba** [comment]
Hi,

thanks for reporting.
It would be helpful to have `--debug` output. 

Also, having so many rules at once make it impossible to find the problem. Are you able to create Github repository with only 1 rule that is causing the problem? Ideally with no other dependencies.	Then I would look into it

**Comment by NotionCommotion** [comment]
Hello Tomas,

The gratitude is all mine and thank you from your effort creating this tool.

I ended up getting `DoctrineAnnotationClassToAttributeRector` to work to some degree, however, was not able to change Doctrine annotations to something like `#[Id, Column(type: "integer"), GeneratedValue(strategy: "IDENTITY")]`.  Assuming just not implemented and maybe not feasible?

I was still not, however, able to get `AnnotationToAttributeRector` and `TypedPropertyRector` to make any changes.  Both are described below and the content is organized as follows:

 - Rule Name
 - rector.php (copied per verbatim from your sample code)
 - Source file (copied per verbatim from your sample code)
 - Output when running `vendor/bin/rector process src/Test --dry-run --debug` on my server
 - Link to getrector.org/demo link
 - Output from getrector.org/demo
 
As seen, `dry-run` doesn't display any changes.

Thanks again!

PS. Consider adding "try it!" links on your documentation page which prepopulate a getrector.org/demo page.
PSS.  Ordered your book the other day and looking forward to receiving it.

### [AnnotationToAttributeRector](https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#annotationtoattributerector)

```
<?php

use Rector\Php80\Rector\Class_\AnnotationToAttributeRector;
use Rector\Php80\ValueObject\AnnotationToAttribute;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Symplify\SymfonyPhpConfig\ValueObjectInliner;

return static function (ContainerConfigurator $containerConfigurator): void {
    $services = $containerConfigurator->services();

    $services->set(AnnotationToAttributeRector::class)
        ->call('configure', [[
            AnnotationToAttributeRector::ANNOTATION_TO_ATTRIBUTE => ValueObjectInliner::inline([
                new AnnotationToAttribute('Symfony\Component\Routing\Annotation\Route', null),
            ]),
        ]]);
};
```
```
<?php

 use Symfony\Component\Routing\Annotation\Route;

 class SymfonyRoute
 {
    /**
     * @Route("/path", name="action")
     */
     public function action()
     {
     }
 }
```

```
[michael@devserver test]$ vendor/bin/rector process src/Test --dry-run --debug
 0/1 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0% < 1 sec/< 1 sec 46.0 MiB[parsing] src/Test/SymfonyRoute.php
[refactoring] src/Test/SymfonyRoute.php
    [applying] Rector\Php80\Rector\Class_\AnnotationToAttributeRector
[post rectors] src/Test/SymfonyRoute.php
[print] src/Test/SymfonyRoute.php


 [OK] Rector is done!


[michael@devserver test]$
```
 
https://getrector.org/demo/1ebdf12f-7d6b-6e66-8acd-f12c34e3b427
 
```
 @@ -1,3 +1,3 @@
 <?php
 
  use Symfony\Component\Routing\Annotation\Route;
@@ -4,10 +4,8 @@
 
  class SymfonyRoute
  {
-    /**
-     * @Route("/path", name="action")
-     */
-     public function action()
-     {
+    #[\Symfony\Component\Routing\Annotation\Route(path: '/path', name: 'action')]
+     public function action()
+     {
      }
- }
+ }
```

### [TypedPropertyRector](https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#typedpropertyrector)

```
<?php
use Rector\Php74\Rector\Property\TypedPropertyRector;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $services = $containerConfigurator->services();

    $services->set(TypedPropertyRector::class)
        ->call('configure', [[
            TypedPropertyRector::CLASS_LIKE_TYPE_ONLY => false,
        ]]);
};
```
```
<?php
final class SomeClass
{
    /**
     * @var int
     */
    private $count;
}
```

```
[michael@devserver test]$ vendor/bin/rector process src/Test --dry-run --debug
 0/2 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0% < 1 sec/< 1 sec 46.0 MiB[parsing] src/Test/DocumentStage.php
[refactoring] src/Test/DocumentStage.php
    [applying] Rector\Php74\Rector\Property\TypedPropertyRector
[post rectors] src/Test/DocumentStage.php
[print] src/Test/DocumentStage.php
[parsing] src/Test/SomeClass.php
[refactoring] src/Test/SomeClass.php
    [applying] Rector\Php74\Rector\Property\TypedPropertyRector
[post rectors] src/Test/SomeClass.php
[print] src/Test/SomeClass.php


 [OK] Rector is done!


[michael@devserver test]$
```

https://getrector.org/demo/1ebdf16e-8774-61ee-a628-2d0e3e8368d6

```
@@ -1,8 +1,5 @@
 <?php
 final class SomeClass
 {
-    /**
-     * @var int
-     */
-    private $count;
+    private int $count;
 }
```

**Comment by TomasVotruba** [comment]
Hi, thanks for more details of your problem


I think we'll need a failing test case to verify this behavior. It's hard to tell from the snippets only.

Here is step by step tutorial how to add it: https://github.com/rectorphp/rector/blob/master/docs/how_to_add_test_for_rector_rule.md

**Comment by NotionCommotion** [comment]
Thanks Tomas for the link.  If it allows me to understand why something is failing or allows others to quickly help, am very happy to create a test.

I am still new to Symfony and have no real experience with PhpUnit (but am willing to learn), and am a little confused on how to create one.  My intent is to first do it with `FinalizeClassesWithoutChildrenRector` per your example before doing more and my `/rector.php` looks like:

```
<?php
declare(strict_types=1);
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Rector\Privatization\Rector\Class_\FinalizeClassesWithoutChildrenRector;
return static function (ContainerConfigurator $containerConfigurator): void {
    $containerConfigurator->services()->set(FinalizeClassesWithoutChildrenRector::class);
};
```
The following is my attempt to follow https://github.com/rectorphp/rector/blob/main/docs/how_to_add_test_for_rector_rule.md.

1. Detect the Rector Rule.  Done.

2. Detect the Minimal File.  Why move it since one could target just the given file anyhow?  Regardless, done.

3. Find the Rector Test Case.  Okay, I understand the naming structure but am getting confused.  Do I _find_ or _create_ this test case?  Is it located in `/vendor/rector` or somewhere else?  I found `/vendor/rector/rector/rules/Privatization/Rector/Class_/FinalizeClassesWithoutChildrenRector.php` and it contains 
 `@see \Rector\Tests\Privatization\Rector\Class_\FinalizeClassesWithoutChildrenRector\FinalizeClassesWithoutChildrenRectorTest` but no `FinalizeClassesWithoutChildrenRector `directory which contains `FinalizeClassesWithoutChildrenRectorTest.php `or a `Fixture` directory.  I have a `/tests` directory which was created when I added phpunit using composer but it only contains `bootstrap.php`. Or should they be located at `/utils/rector/tests/Rector/MyFirstRector/Fixture` as described by [create_own_rule](https://github.com/rectorphp/rector/blob/main/docs/create_own_rule.md
)?

4. Add Change or No-Change Test Fixture File.  Again, no `/Fixture` directory or test fixture file.  Need to understand step 3 before attempting.

Sorry for not understanding better.  Thank you for pointing me in the right direction,.
